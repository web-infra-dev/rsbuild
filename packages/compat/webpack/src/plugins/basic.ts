import path from 'path';
import type { RsbuildPlugin } from '../types';
import { applyBasicPlugin } from '@rsbuild/shared';

/**
 * Provide some basic configs of webpack
 */
export const pluginBasic = (): RsbuildPlugin => ({
  name: 'plugin-basic',

  setup(api) {
    applyBasicPlugin(api);

    api.modifyWebpackChain(async (chain, { env }) => {
      // Disable webpack performance hints.
      // These logs are too complex
      chain.performance.hints(false);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });

      if (env === 'development') {
        // Set correct path for source map
        // this helps VS Code break points working correctly in monorepo
        chain.output.devtoolModuleFilenameTemplate(
          (info: { absoluteResourcePath: string }) =>
            path.resolve(info.absoluteResourcePath).replace(/\\/g, '/'),
        );
      }
    });
  },
});
