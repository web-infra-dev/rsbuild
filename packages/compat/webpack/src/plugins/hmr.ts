import { isUsingHMR, type RsbuildPlugin } from '@rsbuild/shared';

export const pluginHMR = (): RsbuildPlugin => ({
  name: 'rsbuild-webpack:hmr',

  setup(api) {
    api.modifyBundlerChain((chain, utils) => {
      const config = api.getNormalizedConfig();

      if (!isUsingHMR(config, utils)) {
        return;
      }

      const { bundler, CHAIN_ID } = utils;
      chain.plugin(CHAIN_ID.PLUGIN.HMR).use(bundler.HotModuleReplacementPlugin);
    });
  },
});
