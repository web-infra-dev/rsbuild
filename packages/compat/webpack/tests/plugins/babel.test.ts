import { pluginEntry } from '@rsbuild/core/plugins/entry';
import { pluginBabel } from '@/plugins/babel';
import { createStubRsbuild } from '../helper';

describe('plugin-babel', () => {
  it('should set babel-loader', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      rsbuildConfig: {
        output: {
          polyfill: 'entry',
        },
        tools: {
          babel: {},
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should set include/exclude', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      rsbuildConfig: {
        tools: {
          babel(options, { addIncludes, addExcludes }) {
            addIncludes(['src/**/*.ts']);
            addExcludes(['src/**/*.js']);
            return options;
          },
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should apply exclude condition when using source.exclude', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      rsbuildConfig: {
        source: {
          exclude: ['src/foo/**/*.js'],
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should add core-js-entry when output.polyfill is entry', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginBabel()],
      rsbuildConfig: {
        output: {
          polyfill: 'entry',
        },
      },
      entry: {
        main: './index.js',
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();
    expect(config.entry).toMatchSnapshot();
  });

  it('should not add core-js-entry when output.polyfill is usage', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginEntry(), pluginBabel()],
      rsbuildConfig: {
        output: {
          polyfill: 'usage',
        },
      },
      entry: {
        main: './index.js',
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();
    expect(config.entry).toMatchSnapshot();
  });

  it('should override targets of babel-preset-env when using output.overrideBrowserslist config', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      rsbuildConfig: {
        output: {
          overrideBrowserslist: ['Chrome 80'],
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });

  it('should adjust jsescOption config when charset is utf8', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      rsbuildConfig: {
        output: {
          charset: 'utf8',
        },
      },
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(JSON.stringify(config)).toContain(
      '"generatorOpts":{"jsescOption":{"minimal":true}}',
    );
  });

  it('should adjust browserslist when target is node', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginBabel()],
      target: 'node',
    });
    const config = await rsbuild.unwrapWebpackConfig();

    expect(config).toMatchSnapshot();
  });
});
