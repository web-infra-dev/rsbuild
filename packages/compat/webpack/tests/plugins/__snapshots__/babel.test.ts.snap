// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugin-babel > should add core-js-entry when output.polyfill is entry 1`] = `
{
  "main": [
    "/datatext/javascript,import \\"core-js\\";",
    "index.js",
  ],
}
`;

exports[`plugin-babel > should add rule to compile Data URI when enable source.compileJsDataURI 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-babel > should adjust browserslist when target is node 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-dynamic-import-node/lib/index.js",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "node >= 14",
                    ],
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/babel-plugin-dynamic-import-node/lib/index.js",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "node >= 14",
                    ],
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-babel > should apply exclude condition when using source.exclude 1`] = `
{
  "module": {
    "rules": [
      {
        "exclude": [
          "src/foo/**/*.js",
        ],
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-babel > should not add core-js-entry when output.polyfill is usage 1`] = `
{
  "main": [
    "index.js",
  ],
}
`;

exports[`plugin-babel > should override targets of babel-preset-env when using output.overrideBrowserslist config 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "Chrome 80",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "Chrome 80",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-babel > should set babel-loader 1`] = `
{
  "module": {
    "rules": [
      {
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;

exports[`plugin-babel > should set include/exclude 1`] = `
{
  "module": {
    "rules": [
      {
        "exclude": [
          "src/**/*.js",
        ],
        "include": [
          {
            "and": [
              "<ROOT>",
              {
                "not": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/,
              },
            ],
          },
          "src/**/*.ts",
        ],
        "test": /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$\\|\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      {
        "mimetype": {
          "or": [
            "text/javascript",
            "application/javascript",
          ],
        },
        "use": [
          {
            "loader": "<ROOT>/packages/shared/compiled/babel-loader",
            "options": {
              "babelrc": false,
              "compact": false,
              "configFile": false,
              "plugins": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-decorators/lib/index.js",
                  {
                    "version": "legacy",
                  },
                ],
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-export-default-from/lib/index.js",
                "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-partial-application/lib/index.js",
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-proposal-pipeline-operator/lib/index.js",
                  {
                    "proposal": "minimal",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/plugin-transform-runtime/lib/index.js",
                  {
                    "useESModules": true,
                    "version": "7.23.2",
                  },
                ],
                "<ROOT>/packages/babel-preset/dist/pluginLockCorejsVersion",
                [
                  "<ROOT>/packages/compat/webpack/compiled/babel-plugin-lodash",
                  {},
                ],
              ],
              "presets": [
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-env/lib/index.js",
                  {
                    "bugfixes": true,
                    "corejs": {
                      "proposals": true,
                      "version": "3.32",
                    },
                    "exclude": [
                      "transform-typeof-symbol",
                    ],
                    "modules": "commonjs",
                    "targets": [
                      "chrome >= 87",
                      "edge >= 88",
                      "firefox >= 78",
                      "safari >= 14",
                    ],
                    "useBuiltIns": "entry",
                  },
                ],
                [
                  "<ROOT>/node_modules/<PNPM_INNER>/@babel/preset-typescript/lib/index.js",
                  {
                    "allExtensions": true,
                    "allowDeclareFields": true,
                    "allowNamespaces": true,
                    "isTSX": true,
                    "optimizeConstEnums": true,
                  },
                ],
              ],
            },
          },
        ],
      },
    ],
  },
}
`;
